File: 16-3sum-closest/16-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        res = nums[0]+ nums[1] + nums[-1]
        nums.sort()
        
        for i in range(len(nums)):
            l, r = i + 1, len(nums) - 1
            while l < r:
                curr_sum = nums[i] + nums[l] + nums[r]
                if abs(target - curr_sum) < abs(target - res):
                    res = curr_sum
                if curr_sum > target:
                    r -= 1
                else:
                    l += 1
        return res

File: 16-3sum-closest/16-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        //初始一个值，确保可以比较
        Integer res = nums[0] + nums[1] + nums[2];
        for(int i = 0; i < nums.length; i++) {
            int l = i + 1, r = nums.length - 1;
            res = twoSum(nums, i, l, r, target, res);
        }
        return res;
    }
    public int twoSum(int[] nums, int i, int l, int r, int target, Integer res) {
        while(l < r) {
            int sum = nums[i] + nums[l] + nums[r];
            if(Math.abs(sum - target) < Math.abs(res - target)) {
                res = sum;
                // System.out.println("res: " + res);
            }
            if(sum > target) r--;
            else l++;
        }
        return res;
    }
}